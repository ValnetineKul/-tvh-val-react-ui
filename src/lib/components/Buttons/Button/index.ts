import Button from './Button';
import type { ButtonProps, ButtonCommonProps, ButtonIconProps } from './Button';

export default Button;
export type { ButtonProps, ButtonCommonProps, ButtonIconProps };
