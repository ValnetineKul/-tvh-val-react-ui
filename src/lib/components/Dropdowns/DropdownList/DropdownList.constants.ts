export const ACTION_BUTTON_MARGIN = 8;
export const VIRTUALIZED_VISIBLE_ITEMS_COUNT = 6.5;
export const VIRTUALIZED_CONTAINER_PADDING = 0;
export const DEFAULT_ITEM_HEIGHT = 48;
export const LARGE_ITEM_HEIGHT = 64;
export const LIST_CONTAINER_OFFSET = 30;
export const MAX_WINDOW_THRESHOLD = 600;
export const MINIMUM_LIST_HEIGHT = 50;
export const DEFAULT_VISIBLE_ITEMS_COUNT = 100;
