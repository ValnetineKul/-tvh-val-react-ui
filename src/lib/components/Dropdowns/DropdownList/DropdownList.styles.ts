import { makeStyles } from '../../../../themes/core';

type Props = {
  maxListHeight: number;
};

const useStyles = makeStyles<Props>({ name: 'DropdownList' })((theme, props) => ({
  root: {
    borderRadius: theme.borderRadius['borderRadius/none'],
    boxShadow: theme.shadow['shadow/300'],
    display: 'inline-block',
    maxWidth: 320,
    minWidth: 120,
    zIndex: 1292, // TODO Change to theme zIndex value as it will be added.
  },
  surfaceRoot: {
    padding: `${theme.layoutSpacing['spacing/300']} ${theme.layoutSpacing['spacing/000']} ${theme.layoutSpacing['spacing/000']}`,
  },
  header: {
    padding: theme.layoutSpacing['spacing/400'],
  },
  checkbox: {
    padding: `${theme.layoutSpacing['spacing/300']} ${theme.layoutSpacing['spacing/400']}`,
    '& label': {
      width: '100%',
    },
  },
  checkboxLabel: {
    display: 'block',
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
  },
  search: {
    padding: `${theme.layoutSpacing['spacing/300']} ${theme.layoutSpacing['spacing/400']}`,
  },
  virtualList: {
    height: 'unset !important', // TODO is there a way to avoid !important
    maxHeight: props.maxListHeight,
  },
  virtualizedListWrapper: {
    '& > ul': {
      marginBottom: theme.layoutSpacing['spacing/300'],
    },
  },
  list: {
    overflowY: 'auto',
    maxHeight: 312,
    paddingTop: theme.layoutSpacing['spacing/000'],
    paddingBottom: theme.layoutSpacing['spacing/000'],
  },
  listWrapper: {
    paddingBottom: theme.layoutSpacing['spacing/300'],
  },
  dynamicList: {
    maxHeight: props.maxListHeight,
  },
  message: {
    textAlign: 'center',
    padding: theme.layoutSpacing['spacing/400'],
    justifyContent: 'center',
    whiteSpace: 'pre-wrap',
  },
  error: {
    padding: theme.layoutSpacing['spacing/400'],
  },
  fullSpinner: {
    minWidth: 220,
    minHeight: 48,
    display: 'block',
  },
  modalHeightWithSearch: {
    height: '95%',
    top: 'auto',
  },
  modalAutoHeight: {
    height: 'auto',
    maxHeight: '95vh',
    top: 'auto',
  },
  modalTitle: {
    marginLeft: 'auto',
    marginRight: 'auto',
    textOverflow: 'ellipsis',
    overflow: 'hidden',
    whiteSpace: 'nowrap',
  },
  title: {
    padding: `${theme.layoutSpacing['spacing/200']} ${theme.layoutSpacing['spacing/400']}`,
  },
  modalMenuList: {
    maxHeight: '100%',
  },
  modalScrollList: {
    overflowY: 'auto',
    height: '100%',
  },
  modalContentWrapper: {
    display: 'flex',
    flexDirection: 'column',
    height: '100%',
    overflowY: 'auto',
  },
  modalActionItem: {
    marginTop: 'auto',
  },
  actionItem: {
    paddingBottom: theme.layoutSpacing['spacing/300'],
  },
  modalContainer: {
    height: 'auto',
  },
}));

export default useStyles;
