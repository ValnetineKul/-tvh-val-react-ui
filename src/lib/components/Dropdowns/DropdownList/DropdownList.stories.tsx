import { action } from '@storybook/addon-actions';
import React, { useState } from 'react';
import type { StoryTemplate } from '../../../story-utils';
import { createMeta } from '../../../story-utils';
import Banner from '../../Banner';
import Button from '../../Buttons/Button';
import SplitButton from '../../Buttons/SplitButton';
import MenuItem from '../../Menus/MenuItem';
import Typography from '../../Typography';
import DropdownAction from '../DropdownAction';
import DropdownList from './DropdownList';

type Props = React.ComponentProps<typeof DropdownList>;

const Template = ((args: Props) => {
  const [searchValue, setSearchValue] = useState('');
  const [isChecked, setIsChecked] = useState(args.checked);
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);

  const onSearch = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSearchValue(event.target.value);
  };

  const onClose = () => {
    setAnchorEl(null);
    setSearchValue('');
  };

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    anchorEl ? onClose() : setAnchorEl(event.currentTarget);
  };

  const handleChange = () => {
    setIsChecked(!isChecked);
  };

  return (
    <>
      <SplitButton onDropdownClick={handleClick} label="Test button" />
      {anchorEl && (
        <DropdownList
          {...args}
          checked={isChecked}
          listItems={args.listItems.filter(({ props }) =>
            props.label.toLowerCase().includes(searchValue.toLowerCase())
          )}
          anchor={anchorEl}
          onClose={onClose}
          onSearch={onSearch}
          onCheckboxChange={handleChange}
        />
      )}
    </>
  );
}) as StoryTemplate<Props>;

const TemplateEmpty = ((args: Props) => {
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);

  const onClose = () => {
    setAnchorEl(null);
  };

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    anchorEl ? onClose() : setAnchorEl(event.currentTarget);
  };

  return (
    <>
      <SplitButton onDropdownClick={handleClick} label="Test button" />
      {anchorEl && <DropdownList {...args} anchor={anchorEl} onClose={onClose} />}
    </>
  );
}) as StoryTemplate<Props>;

export const Primitive = Template.bind({});
Primitive.args = {
  checkboxLabel: 'Option',
  header: 'Dropdown header',
  inlineMessage: 'No results',
  listItems: new Array(15).fill(null).map((_, idx) => {
    return <MenuItem label={`item_${idx + 1}`} subLabel="testSubLabel" onClick={action('onItemClick')} />;
  }),
  actionItem: <DropdownAction label="label" onClick={action('onActionClick')} />,
  position: 'bottom-end',
  onClose: () => {},
  isLoading: false,
};

Primitive.storyName = 'list';

export const ListWithoutSearchAndCheckbox = TemplateEmpty.bind({});
ListWithoutSearchAndCheckbox.args = {
  header: 'Dropdown header',
  listItems: new Array(15).fill(null).map((_, idx) => {
    return <MenuItem label={`item_${idx + 1}`} onClick={action('onItemClick')} />;
  }),
  position: 'bottom-end',
  onClose: () => {},
  isLoading: false,
};

ListWithoutSearchAndCheckbox.storyName = 'listWithoutSearchAndCheckbox';

export const CustomHeader = Template.bind({});
CustomHeader.args = {
  ...Primitive.args,
  header: (
    <div
      style={{ display: 'flex', gap: 16, justifyContent: 'space-between', alignItems: 'center', marginRight: 'auto' }}
    >
      <Typography variant="h4">Custom Header</Typography>
      <Button label="Button" />
    </div>
  ),
};

CustomHeader.storyName = 'customHeader';

export const TruncatedItem = Template.bind({});
TruncatedItem.args = {
  ...Primitive.args,
  listItems: [
    <MenuItem label="very long menu item name which should be truncated" onClick={action('onItemClick')} />,
    ...Primitive.args.listItems,
  ],
};

TruncatedItem.storyName = 'truncatedItem';

export const EmptyList = TemplateEmpty.bind({});
EmptyList.args = {
  header: 'Dropdown header',
  listItems: [],
  inlineMessage: 'No lists defined,\nplease create a new list.',
  actionItem: <DropdownAction label="label" onClick={action('onActionClick')} />,
  position: 'bottom-end',
  onClose: () => {},
  isLoading: false,
};

EmptyList.storyName = 'emptyList';

export const Error = TemplateEmpty.bind({});
Error.args = {
  header: 'Dropdown header',
  listItems: [],
  error: (
    <Banner
      status="error"
      message="Could not load list. Please refresh."
      button={<Button label="Refresh" onClick={action('onErrorClick')} variant="tertiary" />}
    />
  ),
  actionItem: <DropdownAction label="label" onClick={action('onActionClick')} />,
  position: 'bottom-end',
  onClose: () => {},
  isLoading: false,
};

Error.storyName = 'error';

export default createMeta({
  component: DropdownList,
  title: 'Components/Dropdowns/List',
  argTypes: {
    listItems: {
      control: {
        disable: true,
      },
      table: {
        type: {
          summary: 'An array of MenuItem components',
        },
      },
    },
    error: {
      control: {
        disable: true,
      },
      table: {
        type: {
          summary: 'A component with error message and button',
        },
      },
    },
    header: {
      control: {
        disable: true,
      },
      table: {
        type: {
          summary: 'A string or a component',
        },
      },
    },
    anchor: {
      control: {
        disable: true,
      },
      table: {
        type: {
          summary: 'HTMLElement, used to set the position of the Dropdown',
        },
      },
    },
    actionItem: {
      control: {
        disable: true,
      },
      table: {
        type: {
          summary: 'A component with icon button and label',
        },
      },
    },
    searchField: {
      control: {
        disable: true,
      },
      table: {
        type: {
          summary: 'Search field component',
        },
      },
    },
  },
});
