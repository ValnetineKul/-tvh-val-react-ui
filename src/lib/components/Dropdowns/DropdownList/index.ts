import DropdownList from './DropdownList';

export type { DropdownListProps } from './DropdownList';
export default DropdownList;
