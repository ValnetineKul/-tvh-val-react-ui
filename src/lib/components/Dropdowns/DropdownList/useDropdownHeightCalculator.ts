import React from 'react';
import { MINIMUM_LIST_HEIGHT } from './DropdownList.constants';

const getSpaceAboveAnchor = (anchor: HTMLElement) => {
  return anchor instanceof HTMLElement ? anchor.offsetTop : 0;
};

const getAnchorHeight = (anchor: HTMLElement) => {
  return anchor instanceof HTMLElement ? anchor.offsetHeight : 0;
};

const getSpaceAboveList = (listContainer: HTMLDivElement) => {
  return listContainer instanceof HTMLDivElement ? listContainer.offsetTop : 0;
};

const getActionContainerHeight = (action: HTMLDivElement) => {
  return action instanceof HTMLDivElement ? action.clientHeight : 0;
};

const elementPositionRelativeToWindow = (element: HTMLElement) => {
  const position = ['bottom', 'top'] as const;
  const elementOffset = element instanceof HTMLElement ? element.offsetTop : 0;
  const windowHeight = window.innerHeight;
  return position[Math.round((windowHeight - elementOffset) / windowHeight)];
};

type ReturnType = [
  Map<string, number>,
  (
    anchor: HTMLElement,
    anchorPosition: 'bottom' | 'top',
    listContainer: HTMLDivElement,
    action: HTMLDivElement,
    offset: number
  ) => number
];
const getMaxListHeight = (inCache?: Map<string, number>): ReturnType => {
  const cache = inCache || new Map();

  const saveToCache = (key: string, value: number) => {
    if (value !== 0) {
      cache.set(key, value);
    }
    return value;
  };

  return [
    cache,
    (anchor, anchorPosition, listContainer, action, offset) => {
      let result = 0;
      const spaceAboveList = getSpaceAboveList(listContainer);
      const actionHeight = getActionContainerHeight(action);

      if (anchorPosition === 'top') {
        const spaceAboveAnchor =
          cache.get('spaceAboveAnchor') || saveToCache('spaceAboveAnchor', getSpaceAboveAnchor(anchor));
        const anchorHeight = cache.get('anchorHeight') || saveToCache('anchorHeight', getAnchorHeight(anchor));
        const spaceToDeduct = [spaceAboveAnchor, anchorHeight, spaceAboveList, actionHeight].reduce((a, b) => a + b, 0);

        result = window.innerHeight - spaceToDeduct - offset;
      } else {
        const spaceAboveAnchor = getSpaceAboveAnchor(anchor);
        const spaceToDeduct = [spaceAboveList, actionHeight].reduce((a, b) => a + b, 0);

        result = spaceAboveAnchor - spaceToDeduct - offset;
      }

      return Math.max(MINIMUM_LIST_HEIGHT, result);
    },
  ];
};

function useDropdownHeightCalculator(
  anchor: HTMLElement,
  action: HTMLDivElement,
  visibleItemsCount: number,
  itemHeight: number,
  offset: number
) {
  const [listContainer, setListContainer] = React.useState<HTMLDivElement>(null);
  const [maxListHeight, setMaxListHeight] = React.useState(0);

  const anchorPosition = elementPositionRelativeToWindow(anchor);

  const cache = React.useRef(new Map<string, number>());
  const [returnedCache, getMaxListHeightFn] = getMaxListHeight(cache.current);
  const height = getMaxListHeightFn(anchor, anchorPosition, listContainer, action, offset);

  if (height !== maxListHeight) setMaxListHeight(height);
  cache.current = returnedCache;

  React.useEffect(() => {
    const onResize = () => {
      const maxHeight = getMaxListHeightFn(anchor, anchorPosition, listContainer, action, offset);
      if (maxListHeight === maxHeight) return;
      setMaxListHeight(maxHeight);
    };
    window.addEventListener('resize', onResize);
    return () => {
      window.removeEventListener('resize', onResize);
    };
    // eslint-disable-next-line
  }, []);

  return {
    anchorPosition,
    listHeight: visibleItemsCount * itemHeight,
    maxListHeight,
    setListContainer,
  };
}

export default useDropdownHeightCalculator;
