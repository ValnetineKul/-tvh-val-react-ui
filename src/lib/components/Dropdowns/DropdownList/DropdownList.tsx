import ClickAwayListener from '@mui/material/ClickAwayListener';
import Popper from '@mui/material/Popper';
import type { ChangeEvent, FC } from 'react';
import React, { cloneElement, isValidElement, useState } from 'react';
import { v4 as uuidv4 } from 'uuid';
import useScreenSize from '../../../hooks/useScreenSize';
import MenuList from '../../Menus/MenuList';
import Checkbox from '../../Checkbox';
import Divider from '../../Divider';
import InlineMessage from '../../InlineMessage';
import SearchField from '../../SearchField';
import Typography from '../../Typography';
import Surface from '../../Surfaces/Surface';
import useStyles from './DropdownList.styles';
import SpinnerWithBackdrop from '../../ProgressIndicators/Spinner/SpinnerWithBackdrop';
import Modal from '../../../patterns/Dialogs/Modal';
import type { DataAttributes } from '../../../types/common';
import VirtualizedList from '../../VirtualizedList';
import useDropdownHeightCalculator from './useDropdownHeightCalculator';
import type { MenuItemProps } from '../../Menus/MenuItem/MenuItem.types';
import {
  ACTION_BUTTON_MARGIN,
  DEFAULT_ITEM_HEIGHT,
  MAX_WINDOW_THRESHOLD,
  VIRTUALIZED_VISIBLE_ITEMS_COUNT,
  LARGE_ITEM_HEIGHT,
  LIST_CONTAINER_OFFSET,
  VIRTUALIZED_CONTAINER_PADDING,
} from './DropdownList.constants';

function getListItemHeight<T extends { props: MenuItemProps }[]>(items: T) {
  const itemData = React.Children.toArray(items) as T;

  return itemData[0].props.subLabel ? LARGE_ITEM_HEIGHT : DEFAULT_ITEM_HEIGHT;
}

export interface DropdownListProps {
  anchor?: HTMLElement | null;
  listItems: React.ReactElement[];
  header?: string | React.ReactNode;
  inlineMessage?: string;
  error?: React.ReactNode;
  actionItem?: React.ReactNode;
  searchPlaceholder?: string;
  searchAutoFocus?: boolean;
  checkboxLabel?: string;
  checked?: boolean;
  isLoading?: boolean;
  isVirtualized?: boolean;
  visibleItemsCount?: number;
  position?: 'bottom-end' | 'bottom-start' | 'bottom' | 'top-end' | 'top-start' | 'top';
  menuListClassName?: string;
  onClose: () => void;
  onSearch?: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onCheckboxChange?: (event: React.ChangeEvent<HTMLInputElement>, checked: boolean) => void;
  searchFieldProps?: React.HTMLAttributes<HTMLInputElement | HTMLTextAreaElement> & DataAttributes;
  checkboxFieldProps?: React.HTMLAttributes<HTMLInputElement> & DataAttributes;
  clearButtonProps?: React.HTMLAttributes<HTMLButtonElement> & DataAttributes;
}

const DropdownList: FC<DropdownListProps> = ({
  anchor,
  listItems,
  header,
  inlineMessage,
  error,
  actionItem,
  searchPlaceholder,
  searchAutoFocus = false,
  checkboxLabel,
  checked,
  isLoading,
  isVirtualized = true,
  visibleItemsCount: visibleItemsCountProp,
  position,
  menuListClassName,
  onClose,
  onSearch,
  onCheckboxChange,
  searchFieldProps,
  clearButtonProps,
  checkboxFieldProps,
}) => {
  const [listHeight, setListHeight] = useState(0);
  const actionRef = React.useRef<HTMLDivElement>(null);

  const visibleItemsCount = visibleItemsCountProp || VIRTUALIZED_VISIBLE_ITEMS_COUNT;
  const {
    anchorPosition,
    listHeight: estimatedListHeight,
    maxListHeight,
    setListContainer,
  } = useDropdownHeightCalculator(
    anchor?.offsetParent as HTMLElement,
    actionRef.current,
    visibleItemsCount,
    listItems.length > 0 ? getListItemHeight(listItems) : 0,
    LIST_CONTAINER_OFFSET + (actionRef.current ? ACTION_BUTTON_MARGIN : 0)
  );
  const { classes, cx } = useStyles({ maxListHeight });

  const { isTabletUp } = useScreenSize('Tablet');
  const isMobile = !isTabletUp;

  const shouldShowMaxListItems =
    isTabletUp &&
    ((window.innerHeight <= MAX_WINDOW_THRESHOLD && !visibleItemsCountProp) ||
      (!!visibleItemsCountProp && estimatedListHeight > maxListHeight));

  const handleResize = (entries: ResizeObserverEntry[]) => {
    if (!isMobile || !isVirtualized) return;
    const entry = entries[0];
    let { height } = entry.contentRect;
    if (actionRef.current) height -= actionRef.current.clientHeight + ACTION_BUTTON_MARGIN;

    if (height && height !== listHeight && !isNaN(height)) setListHeight(height);
  };

  const resizeObserver = new ResizeObserver(handleResize);

  React.useEffect(() => {
    return () => {
      resizeObserver.disconnect();
    };
    // eslint-disable-next-line
  }, [isVirtualized, isMobile]);

  const onRefChange = React.useCallback(
    (listRef) => {
      if (!isMobile || !isVirtualized || !listRef) return;
      resizeObserver.observe(listRef);
    },
    // eslint-disable-next-line
    [isVirtualized, isMobile]
  );

  const handleCloseList = (event: MouseEvent | TouchEvent) => {
    if (anchor && anchor.contains(event.target as HTMLElement)) {
      return;
    }
    onClose();
  };

  const [searchValue, setSearchValue] = React.useState('');

  const handleSearchChange = (e: ChangeEvent<HTMLInputElement>) => {
    setSearchValue(e.target.value);
    onSearch && onSearch(e);
  };

  const renderHeader = () => {
    if (header) {
      return (
        <>
          <div className={classes.header}>
            {typeof header === 'string' ? <Typography variant="h4">{header}</Typography> : header}
          </div>
          <Divider />
        </>
      );
    }
    return null;
  };

  const renderMobileHeader = () => {
    if (header) {
      return (
        <>
          {typeof header === 'string' ? (
            <Typography variant="h5" className={classes.modalTitle} component="span">
              {header}
            </Typography>
          ) : (
            header
          )}
        </>
      );
    }
    return null;
  };

  const renderCheckbox = () => {
    if (onCheckboxChange && checkboxLabel) {
      return (
        <>
          <div className={classes.checkbox}>
            <Checkbox
              inputProps={checkboxFieldProps}
              onChange={onCheckboxChange}
              label={checkboxLabel}
              checked={checked}
              className={classes.checkboxLabel}
            />
          </div>
          {isMobile && !onSearch && <Divider />}
        </>
      );
    }
    return null;
  };

  const renderSearchField = () => {
    if (onSearch) {
      return (
        <div className={classes.search}>
          <SearchField
            searchFieldProps={searchFieldProps}
            deleteIconProps={clearButtonProps}
            fullWidth
            value={searchValue}
            onChange={handleSearchChange}
            placeholder={searchPlaceholder}
            autoFocus={searchAutoFocus}
            variant="secondary"
          />
        </div>
      );
    }
    return null;
  };

  const renderItems = () => {
    if (error) {
      return <div className={classes.error}>{error}</div>;
    }

    if (listItems?.length > 0) {
      const listItemsWithKeys = listItems.map((item, key) =>
        isValidElement(item) ? cloneElement(item, { key }) : item
      );
      if (isVirtualized) {
        return (
          <VirtualizedList
            className={cx({
              [classes.virtualList]: shouldShowMaxListItems,
              [classes.virtualizedListWrapper]: !actionItem,
            })}
            listboxClassName={cx(classes.list, menuListClassName, {
              [classes.modalMenuList]: isMobile,
            })}
            visibleItemsCount={visibleItemsCount}
            containerPadding={VIRTUALIZED_CONTAINER_PADDING}
            mobileListheight={listHeight}
          >
            {listItemsWithKeys}
          </VirtualizedList>
        );
      }

      return (
        <MenuList
          className={cx(classes.list, { [classes.dynamicList]: shouldShowMaxListItems }, menuListClassName, {
            [classes.modalMenuList]: isMobile,
            [classes.listWrapper]: !actionItem,
          })}
        >
          {listItemsWithKeys}
        </MenuList>
      );
    }

    if (inlineMessage) {
      return <InlineMessage className={classes.message} message={inlineMessage} />;
    }

    return null;
  };

  const renderSpinnerWithBackdrop = () => {
    return (
      <SpinnerWithBackdrop
        ref={setListContainer}
        isBackdropOpaque={!listItems.length}
        className={cx(classes.fullSpinner, { [classes.modalScrollList]: isMobile })}
        size="lg"
        isLoading={isLoading}
      >
        {renderItems()}
      </SpinnerWithBackdrop>
    );
  };

  const renderActionItem = () => {
    if (actionItem) {
      return (
        <div ref={actionRef} className={cx(classes.actionItem, { [classes.modalActionItem]: isMobile })}>
          <Divider />
          {actionItem}
        </div>
      );
    }
    return null;
  };

  if (isTabletUp) {
    return (
      <ClickAwayListener onClickAway={handleCloseList}>
        <Popper
          className={classes.root}
          open={!!anchor}
          anchorEl={anchor}
          placement={position || (anchorPosition === 'top' ? 'bottom-end' : 'top-end')}
          modifiers={[
            {
              name: 'offset',
              options: {
                offset: [0, 4],
              },
            },
          ]}
        >
          <Surface className={classes.surfaceRoot} color="100">
            {renderHeader()}
            {renderCheckbox()}
            {renderSearchField()}
            {renderSpinnerWithBackdrop()}
            {renderActionItem()}
          </Surface>
        </Popper>
      </ClickAwayListener>
    );
  }

  return (
    <Modal
      open={!!anchor}
      id={typeof header === 'string' ? header : uuidv4()}
      title={renderMobileHeader()}
      onClose={onClose}
      rootClassName={onSearch || isVirtualized ? classes.modalHeightWithSearch : classes.modalAutoHeight}
      containerClassName={classes.modalContainer}
      titleClassName={classes.title}
    >
      {renderCheckbox()}
      {renderSearchField()}
      <div ref={onRefChange} className={classes.modalContentWrapper}>
        {renderSpinnerWithBackdrop()}
        {renderActionItem()}
      </div>
    </Modal>
  );
};

export default DropdownList;
