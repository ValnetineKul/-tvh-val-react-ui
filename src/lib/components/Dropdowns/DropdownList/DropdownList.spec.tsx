import React from 'react';
import { render, screen, userEvent, initResponsiveTest } from '../../../test-utils';

import DropdownList from './DropdownList';
import MenuItem from '../../Menus/MenuItem';
import Button from '../../Buttons/Button';

const props = {
  anchor: document.createElement('div'),
  isOpen: true,
  listItems: [<MenuItem label="item_1" onClick={jest.fn()} />, <MenuItem label="item_2" onClick={jest.fn()} />],
  onClose: jest.fn(),
  header: 'Dropdown header',
  actionItem: <div>Action</div>,
};

describe('DropdownList', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  beforeAll(() => {
    jest.useFakeTimers();
  });
  afterAll(() => {
    jest.useRealTimers();
  });

  it('Should render menu items', () => {
    initResponsiveTest('Desktop');
    render(<DropdownList {...props} />);
    const dropdown = screen.getByRole('tooltip');
    expect(dropdown).toMatchSnapshot();
  });

  it('Should be virtualized', () => {
    initResponsiveTest('Desktop');
    const manyItems = Array.from(new Array(20)).map((_, index) => {
      return <MenuItem label="item" key={index} onClick={jest.fn()} />;
    });
    render(<DropdownList {...props} listItems={manyItems} />);
    const items = screen.getAllByText('item');
    expect(items).toHaveLength(12);
  });

  it('Should not be virtualized', () => {
    const manyItems = Array.from(new Array(20)).map((_, index) => {
      return <MenuItem label="item" key={index} onClick={jest.fn()} />;
    });
    render(<DropdownList {...props} listItems={manyItems} isVirtualized={false} />);
    const items = screen.getAllByText('item');
    expect(items).toHaveLength(20);
  });

  it('Should render inline message if no list items', () => {
    render(<DropdownList {...props} listItems={[]} inlineMessage="message" />);
    const inlineMessage = screen.getByText('message');
    expect(inlineMessage).toBeInTheDocument();
  });

  it('Should render error banner', () => {
    render(<DropdownList {...props} error={<div>Error message</div>} />);
    const error = screen.getByText('Error message');
    expect(error).toBeInTheDocument();
  });

  it('Should call onClose when click outside', () => {
    initResponsiveTest('Desktop');
    const onClose = jest.fn();
    render(<DropdownList {...props} onClose={onClose} />);
    jest.advanceTimersByTime(0);
    userEvent.click(document.body);
    expect(onClose).toHaveBeenCalledTimes(1);
  });

  it('Should render Spinner when it is a loading state', () => {
    render(<DropdownList {...props} anchor={document.body} isLoading />);
    const spinner = screen.getByRole('progressbar');
    expect(spinner).toBeInTheDocument();
  });

  it('Should have overwritten maxHeight style', () => {
    const { getByRole } = render(<DropdownList {...props} menuListClassName="someClassName" isVirtualized={false} />);
    expect((getByRole('menu') as HTMLElement).className).toMatch('someClassName');
  });

  it('Should render custom header', () => {
    const args = {
      ...props,
      header: <Button label="Custom header" />,
    };
    render(<DropdownList {...args} />);
    const header = screen.getByRole('button', { name: 'Custom header' });
    expect(header).toBeInTheDocument();
  });

  describe('Should set correct component to display menu items', () => {
    it('Should show modal on mobile', () => {
      initResponsiveTest('Mobile');

      render(<DropdownList {...props} />);
      expect(screen.getByRole('dialog')).toBeInTheDocument();
      expect(screen.queryByRole('tooltip')).not.toBeInTheDocument();
    });

    it('Should show tooltip on tablet', () => {
      initResponsiveTest('Tablet');

      render(<DropdownList {...props} />);
      expect(screen.getByRole('tooltip')).toBeInTheDocument();
      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    });

    it('Should show tooltip on desktop', () => {
      initResponsiveTest('Desktop');

      render(<DropdownList {...props} />);
      expect(screen.getByRole('tooltip')).toBeInTheDocument();
      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    });
  });
});
