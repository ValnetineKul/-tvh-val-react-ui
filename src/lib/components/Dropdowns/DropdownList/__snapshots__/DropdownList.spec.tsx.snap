// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownList Should render menu items 1`] = `
<div
  class="tss-194ixmc-DropdownList-root"
  data-popper-escaped=""
  data-popper-placement="bottom-end"
  data-popper-reference-hidden=""
  role="tooltip"
  style="position: absolute; left: 0px; top: 0px; margin: 0px; right: 0px; transform: translate(0px, 4px);"
>
  <div
    class="tss-wkvj97-Surface-root-DropdownList-surfaceRoot"
  >
    <div
      class="tss-vja05h-DropdownList-header"
    >
      <h4
        class="tss-1-Typography-headerFunctional-ref tss-l9ejq1-Typography-root-Typography-headerFunctional-Typography-h4"
      >
        Dropdown header
      </h4>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth tss-1pam8iy-Divider-root-Divider-horizontal mui-9mgopn-MuiDivider-root"
    />
    <div
      class="tss-1eox161-SpinnerWithBackdrop-root-DropdownList-fullSpinner"
    >
      <div
        class=""
      >
        <div
          class=""
        >
          <div
            class="tss-3u1d75-VirtualizedList-autosizer"
            style="overflow: visible;"
          >
            <div
              class="tss-182o6y1-VirtualizedList-root"
              style="position: relative; height: 96px; width: 100%; overflow: auto; will-change: transform; direction: ltr;"
            >
              <ul
                style="height: 96px; width: 100%;"
              >
                <li
                  class="tss-1m62q20-VirtualizedList-row"
                  style="position: absolute; left: 0px; top: 0px; height: 48px; max-width: 100%;"
                >
                  <div
                    class="tss-1awyf5s-MenuItem-root MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root  mui-t7a2rv-MuiButtonBase-root-MuiMenuItem-root"
                    role="menuitem"
                    tabindex="-1"
                  >
                    <button
                      class="MuiButtonBase-root tss-22-MenuItem-wrapper-ref tss-19jpft6-ButtonBase-root-ButtonBase-focusInset-MenuItem-wrapper mui-10d1a0h-MuiButtonBase-root"
                      tabindex="0"
                      type="button"
                    >
                      <div
                        class="tss-sqs6xo-MenuItem-labelWrapper-MenuItem-label"
                      >
                        <div
                          class="tss-1-Typography-weightRegular-ref tss-m39tug-Typography-root-Typography-weightRegular-Tooltip-wrapper-TruncatedText-truncatedText-MenuItem-label-Typography-body400"
                          data-mui-internal-clone-element="true"
                        >
                          <span
                            class="tss-1-Typography-weightRegular-ref tss-b5vov3-Typography-root-Typography-weightRegular-TruncatedText-truncatedText-Typography-body500"
                          >
                            item_1
                          </span>
                        </div>
                      </div>
                    </button>
                  </div>
                </li>
                <li
                  class="tss-1m62q20-VirtualizedList-row"
                  style="position: absolute; left: 0px; top: 48px; height: 48px; max-width: 100%;"
                >
                  <div
                    class="tss-1awyf5s-MenuItem-root MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root  mui-t7a2rv-MuiButtonBase-root-MuiMenuItem-root"
                    role="menuitem"
                    tabindex="-1"
                  >
                    <button
                      class="MuiButtonBase-root tss-22-MenuItem-wrapper-ref tss-19jpft6-ButtonBase-root-ButtonBase-focusInset-MenuItem-wrapper mui-10d1a0h-MuiButtonBase-root"
                      tabindex="0"
                      type="button"
                    >
                      <div
                        class="tss-sqs6xo-MenuItem-labelWrapper-MenuItem-label"
                      >
                        <div
                          class="tss-1-Typography-weightRegular-ref tss-m39tug-Typography-root-Typography-weightRegular-Tooltip-wrapper-TruncatedText-truncatedText-MenuItem-label-Typography-body400"
                          data-mui-internal-clone-element="true"
                        >
                          <span
                            class="tss-1-Typography-weightRegular-ref tss-b5vov3-Typography-root-Typography-weightRegular-TruncatedText-truncatedText-Typography-body500"
                          >
                            item_2
                          </span>
                        </div>
                      </div>
                    </button>
                  </div>
                </li>
              </ul>
            </div>
          </div>
          <div
            class="resize-triggers"
          >
            <div
              class="expand-trigger"
            >
              <div
                style="width: 1px; height: 101px;"
              />
            </div>
            <div
              class="contract-trigger"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="tss-zwovav-DropdownList-actionItem"
    >
      <hr
        class="MuiDivider-root MuiDivider-fullWidth tss-1pam8iy-Divider-root-Divider-horizontal mui-9mgopn-MuiDivider-root"
      />
      <div>
        Action
      </div>
    </div>
  </div>
</div>
`;
